ipso-syntax/src/desugar/mod.rs

```diff
@@ -123,6 +122,347 @@ fn desugar_cmd_part_mut(source: &Source, cmd_part: &mut CmdPart) -> Result<(), E
+    The type checker expects that there are no duplicate variant constructors in a `case`
+    expression.
+
+    Example:
+
+    ```
+    case x of
+      A 1 -> "A 1"
+      B _ -> "B _"
+      A _ -> "A _"
+    ```
+
+    Without reordering, the above code desugars to:
+
+    ```
+    f : (| A : Int, B : Int |) -> String
+    f x =
+      case x of
+        A fresh_1 ->
+          case fresh_1 of
+            1 -> "A 1"
+        B fresh_2 ->
+          case fresh_2 of
+            _ -> "B _"
+        A fresh_3 ->
+          case fresh_3 of
+            _ -> "A _"
+    ```
+
+    The type checker considers the `A fresh_3` pattern redundant.
+
+    Instead, we first reorder the `A` constructors to reside next to each other:
+
+    ```
+    case x of
+      A 1 -> "A 1"
+      A _ -> "A _"
+      B _ -> "B _"
+    ```
+
+    And then desugar:
+
+    ```
+    f : (| A : Int, B : Int |) -> String
+    f x =
+      case x of
+        A fresh_1 ->
+          case fresh_1 of
+            1 -> "A 1"
+            _ -> "A _"
+        B fresh_2 ->
+            case fresh_2 of
+            _ -> "B _"
+    ```
+
+    Resulting in a `case` expression with no redundant matches.
+    */
+    fn reorder_branches(branches: Vec<Branch>) -> Vec<Branch> {
+        struct Branches {
+            first: Branch,
+            rest: Vec<Branch>,
+        }
+
+        let mut constructor_indices: HashMap<Rc<str>, usize> = HashMap::new();
+        let mut grouped_branches: Vec<Branches> = Vec::new();
+        for branch in branches {
+            match &branch.pattern.item {
+                Pattern::Variant { name, .. } => {
+                    let name = name.clone();
+                    match constructor_indices.get(&name) {
+                        Some(index) => {
+                            grouped_branches.get_mut(*index).unwrap().rest.push(branch);
+                        }
+                        None => {
+                            let index = grouped_branches.len();
+                            grouped_branches.push(Branches {
+                                first: branch,
+                                rest: Vec::new(),
+                            });
+                            constructor_indices.insert(name, index);
+                        }
+                    };
+                }
+                Pattern::Name(_)
+                | Pattern::Record { .. }
+                | Pattern::Char(_)
+                | Pattern::Int(_)
+                | Pattern::String(_)
+                | Pattern::Wildcard => grouped_branches.push(Branches {
+                    first: branch,
+                    rest: Vec::new(),
+                }),
+            }
+        }
+
+        grouped_branches
+            .into_iter()
+            .flat_map(|group| std::iter::once(group.first).chain(group.rest.into_iter()))
+            .collect()
+    }
+
+    enum Grouped {
```

Comment:
I think I can reuse `reorder_branches::Branches`. It sort of looks like I'm grouping the branches twice.